hbase-mapreduce/src/main/java/org/apache/hadoop/hbase/mob/mapreduce/MobRefReporter.java

```diff
@@ -0,0 +1,485 @@
+ * that the MOB system will look when attempting to resolve the actual value.
+ *
+ * The job includes counters that can help provide a rough sketch of the mob data.
+ *
+ * <pre>
+ * Map-Reduce Framework
+ *         Map input records=10000
+ * ...
+ *         Reduce output records=99
+ * ...
+ * CELLS_PER_ROW_DIGITS
+ *         1=10000
+ * MOB
+ *         NUM_CELLS=52364
+ * PROBLEM
+ *         IMPACTED_ROWS=338
+ *         MOB_FILES=2
+ * PROBLEM_ROWS_PER_FILE_DIGITS
+ *         3=2
+ * SIZE_PER_CELL_DIGITS
+ *         5=627
+ *         6=51392
+ *         7=345
+ * SIZE_PER_ROW_DIGITS
+ *         6=6838
+ *         7=3162
+ * </pre>
+ *
+ *   * Map-Reduce Framework:Map input records - the number of rows with mob references
+ *   * Map-Reduce Framework:Reduce output records - the number of unique hfiles referenced
+ *   * MOB:NUM_CELLS - the total number of mob reference cells
+ *   * PROBLEM:IMPACTED_ROWS - the number of rows that reference hfiles with an issue
+ *   * PROBLEM:MOB_FILES - the number of unique hfiles that have an issue
+ *   * CELLS_PER_ROW_DIGITS: - this counter group gives a histogram of the order of magnitude of the
+ *         number of cells in a given row by grouping by the number of digits used in each count.
+ *         This allows us to see more about the distribution of cells than what we can determine
+ *         with just the cell count and the row count. In this particular example we can see that
+ *         all of our rows have somewhere between 1 - 9 cells.
+ *   * PROBLEM_ROWS_PER_FILE_DIGITS: - this counter group gives a histogram of the order of
+ *         magnitude of the number of rows in each of the hfiles with a problem. e.g. in the
+ *         example there are 2 hfiles and they each have the same order of magnitude number of rows,
+ *         namely 3 digits worth i.e. 100-999.
+ *   * SIZE_PER_CELL_DIGITS: - this counter group gives a histogram of the order of magnitude of
+ *         the size of mob values according to our reference cells. e.g. in the example above we
+ *         have between 5 and 7 digits in our cell sizes i.e. 10,000 - 9,999,999 bytes. From this
+ *         histogram we can also see that _most_ cells are 100,000 - 199,000 bytes and the smaller
+ *         and bigger ones are outliers making up less than 2% of mob cells.
+ *   * SIZE_PER_ROW_DIGITS: - this counter group gives a histogram of the order of magnitude of the
+ *         size of
+ *
+ * Generates a report that gives one file status per line, with tabs dividing fields.
+ *
+ * <pre>
+ * RESULT OF LOOKUP	FILE REF	comma seperated, base64 encoded rows when there's a problem
+ * </pre>
+ *
+ * e.g.
+ *
+ * <pre>
+ * MOB DIR	09c576e28a65ed2ead0004d192ffaa382019110184b30a1c7e034573bf8580aef8393402
+ * MISSING FILE    28e252d7f013973174750d483d358fa020191101f73536e7133f4cd3ab1065edf588d509        MmJiMjMyYzBiMTNjNzc0OTY1ZWY4NTU4ZjBmYmQ2MTUtNTIz,MmEzOGE0YTkzMTZjNDllNWE4MzM1MTdjNDVkMzEwNzAtODg=
```

Comment:
base64 can be used in lots of places and the hex output from the hbase shell can basically just be used there.

Would examples for using the base64 output in the hbase shell help? I have some handy.

Reply:
I would likely simply use `Base64.encode()`/`Base64.decode()` in `hbase shell` if that is what you are suggesting?

For me in my simple operations, I usually think of regions when coming across a base64 encoded string from [HRegionInfo.encode()](https://hbase.apache.org/1.2/apidocs/org/apache/hadoop/hbase/HRegionInfo.html#encodeRegionName(byte[])). When seeing a row, I think of a hex escaped ASCII representation ala HBase shell. There may not be consistency across tools though or perhaps now there is and I am ignorantly living in HBase 1.x vs HBase 2.x and master?

Reply:
It is sadly still not consistent anywhere.

The thing for me is that as a base64 encoded string I'm more likely to be able to use a single set of instructions when helping someone else (vs different when it might be hex and might not) and I'm much more likely to notice if the row key gets mangled somewhere between the report running, stuff getting copied into an email or a support ticketing system somewhere, and eventually making its way to me. (Or to a mailing list)

Using `Base64.decode` in the hbase shell is exactly correct and what my examples all do.

Reply:
Interesting. I like your reasons. I only appreciate I can see if it is likely contiguous regions with ASCII keys and similarly rationalize about issues otherwise reported. I would say this question begs the code making things more directed to one solution by design but I am happy either way.

Reply:
👍 if it turns out we need the "what are the regions impacted" I could make a simple follow on to get that from the set of rows. I suspect I can do that with a loop in the shell.